packages/file/src/managers/directory.manager.ts

```diff
@@ -0,0 +1,163 @@
+import "reflect-metadata"
+import {injectable} from "tsyringe";
+import {DirectoryListOptions} from "../options/directory-list.options";
+import {FileInfoInterface} from "../interfaces/file-info.interface";
+import {MatchTypeEnum} from "../enums/match-type.enum";
+import {TypesEnum} from "../enums/types.enum";
+import {DirectoryListResultEnum} from "../enums/directory-list-result.enum";
+import {promises as fsp} from "fs";
+import path from "path";
+
+@injectable()
+export class DirectoryManager {
+    private async listRecursively(dir: string, options: DirectoryListOptions) {
+
+        // collect sub-directories for possible recursion
+        const dirs = [];
+
+        const files = await fsp.readdir(dir, {withFileTypes: true});
+        for (const dirEnt of files) {
+            const filename = dirEnt.name;
+            const fullPath = path.join(dir, filename);
+            const ext = path.extname(filename).slice(1);                  // file extension without leading "."
+            let basename = filename;
+            if (ext.length) {
+                basename = filename.slice(0, -(ext.length + 1));          // get part without the extension
+            }
+
+            // put all this in an object for later use
+            const obj: FileInfoInterface = {filename, basename, extension: ext, fullPath, type: ""};
+
+            // get type of entry
+            let type = "<unknown>";
+            let include = false;
+            if (dirEnt.isFile()) {
+                type = "file";
+                include = options.types === TypesEnum.File || options.types === TypesEnum.ALL;
+            } else if (dirEnt.isDirectory()) {
+                type = "dir";
+                include = options.types === TypesEnum.Directory || options.types === TypesEnum.ALL;
+                dirs.push(obj);             // save dir for recursion option
+            }
+            obj.type = type;
+
+            if (!include) {
+                continue;           // skip further processing for this file/dir
+            }
+
+            // see if we should skip top level files
+            if (options.skipTopLevelFiles && type === "file") {
+                continue;
+            }
+
+            // if doing matching
+            if (options.match) {
+                let target;
+                if (options.matchType === MatchTypeEnum.Extension) {
+                    target = ext;
+                } else if (options.matchType === MatchTypeEnum.Base) {
+                    target = basename;
+                } else if (options.matchType === MatchTypeEnum.Filename) {
+                    target = filename;
+                } else {
+                    throw new Error(`options.matchType contains invalid value "${options.matchType}", should be "MatchTypeEnum.Extension", "MatchTypeEnum.Base" or "MatchTypeEnum.Filename"`);
```

Comment:
Here I'd add some context to the error message to make it clear it's happened during listing the files